### PR TITLE
Add support to action run command to inherit all the environment variables

### PR DESCRIPTION
**Pull request by Kami**
This pull request allows user to easily send all the environment variables which are accessible to the CLI as ``env`` parameter to the action being executed.

Keep in mind that some common environment variables which are set in a login shell such as ``PWD``, ``PATH`` and others are ignored. The feature is there to make it easier for user to send user-defined environment variables (e.g. different credentials, path to credential files, etc.) and not the ones which are defined by login script / shell. On top of that, sending some of those variables (e.g. ``PATH``, ``PYTHONPATH``, etc.) could interfere and break the action itself.

See #1034 for more information.

### CODE REVIEW

**Comment by manasdk** [comment]
I like the direction but wouldn't it be better to provide a white list of env vars as part of the command. In some sense  that makes the result a lot more predictable.

so something like,

--inherit-env OS_USERNAME, OS_PASSWORD...

**Comment by Kami** [comment]
@manasdk I was thinking about it and we could do that, but then we don't really save much typing since you could also do `--env="{\"OS_USERNAME\": \"$OS_USERNAME\",.. \}`.

What do other people think? /cc @jtopjian

**Comment by manasdk** [comment]
Does --env will cause the value to be picked from local shell?

**Comment by Kami** [comment]
@manasdk with --env you need to specify key and values, which means you can also use existing environment variables for the values (you just need to be careful how you quote / escape the variables since in this case you wanted them interpolated locally and not on the remote server).

**Comment by manasdk** [comment]
> you can also use existing environment variables for the values

The cli somehow tries to do a lookup of the environment var locally?

**Comment by Kami** [comment]
Not the CLI, the shell does though. Ddepending on how you quote / escape the variables, they will or won't be interpolated locally - http://docs.stackstorm.com/latest/cli.html#escaping-shell-variables-when-using-core-local-and-core-remote-actions

**Comment by manasdk** [comment]
sorry I did not see $OS_USERNAME and which is why I got confused. makes sense.

Anyway, my original point was that --inhertit-env can lead to surprising results but this works since the failure would be obvious due to error and introspection of the created execution.

**Comment by Kami** [comment]
Honestly, I also prefer being explicit, so if @jtopjian is OK with it I can change it to ``--inherit-envs=MY_VAR_1,MY_VAR_2,...`

**Comment by jtopjian** [comment]
I understand the position you guys are coming from, but to be honest, `--inherit-envs` is just sort of pushing food around the plate. :smile: 

What about:

* Only accepting `ST2_*` variables
* Specifying `--env-file=foorc`

**Comment by Kami** [comment]
In this case it's probably better to just stick with the current approach.

And if we see that people find it too confusing, we can change it later :)

**Comment by jtopjian** [comment]
So to confirm: `--inherit-env` has been merged and is "live"? If so, I'll make some time to test this out -- hopefully over the weekend.

**Comment by Kami** [comment]
@jtopjian It's only been merged into master so far which means it will only work if you are using source install method and you are running off latest master.

Besides that, it will be included in the next release (0.8.0), but we don't have ETA for it yet.

**Comment by DoriftoShoes** [comment]
@kami we build rpm and deb packages for top of the tree also. They are available via the same download methods as the stable branch.

**Comment by Kami** [comment]
@DoriftoShoes Thanks, good point. 

@jtopjian as @DoriftoShoes pointed out, if you want to play with the latest development version you can also use packages from https://ops.stackstorm.net/releases/st2/0.8dev/

**Comment by dzimine** [comment]
or run `st2_deploy.sh 0.8dev` when you follow the installation. Same re updating st2express for Vagrant or docker.

Ok we need to put this bit of info in http://docs.stackstorm.com/install/index.html#installation

**Comment by dzimine** [comment]
@jtopjian how would you want to handle your existing files with env variables? Should we give an ability to read a file? 

Let's move the further conversation to #1034 and close that bug only when @DoriftoShoes and @jtopjian  happy :)

**Comment by jtopjian** [comment]
hm, I might have spoke too soon in #1034. (jumping back here since I think this is a bug).

I've been playing around with this as well as building some actions. I ended up opening a new ssh session and running commands expecting them to fail, but sure enough, they were succeeding. It turns out they had access to the environment variables passed in by "session 1".

It wasn't until I logged out of "session 1" and restarted st2 (`st2ctl restart`) that everything flushed. Now I'm unable to reproduce the problem... 

Any idea how that could have happened?

**Comment by Kami** [comment]
@jtopjian  I assume this was a core.remote action (or was it core.local)?

Also, what kind if command did you run? Can you paste the action file so we can try to reproduce it?

**Comment by jtopjian** [comment]
I've tried to reproduce this ever since I reported the issue and had no luck. Here are the actions I was trying to do:

* Get the openstack pack to work by modifying the Python code
* Modified one of the commands to a Python runner
* Did not set an exit 0 at the end
* Added an exit 0 at the end

Around here is where I opened a new session and noticed it had access to the environment

* Restarted st2
* Reloaded the pack

Nothing worked until I logged out the first user.

**Comment by Kami** [comment]
Thanks for more details.

I will look into it and try to reproduce it as well. Might be some weird edge case with shell sessions (those are only used with local runner and in some cases with remote runner though).

**Comment by jtopjian** [comment]
@Kami just a quick note to say that I figured out how to reproduce the above issue:

When starting services via `st2ctl`, the st2 components will inherit all environment variables of the logged in user who issued that command.

I figured this out with the new OpenStack pack that utilizes environment variables. If I'm logged as a user, source an openrc file, then restart st2 via `st2ctl`, then i'm able to run any openstack action and it'll work without having to specify any credentials since the OS_* variables are part of the st2 procs' environment.

File: st2client/st2client/commands/action.py [hunk 1]
```diff
@@ -424,6 +456,17 @@ def _format_error_result(self, result):
                 result['traceback'])
         return result
 
+    def _get_inherited_env_vars(self):
```
Comment (Kami) [inline st2client/st2client/commands/action.py]: @lakshmi-kannan 

> lakstorm [03:18]
So it appears we make a copy of the env variables when the CLI is executed. This is then available to actions.
lakstorm [03:18]
So confused why we did this.
lakstorm [03:19]
so even though auth_token is emtpy, in 0.12, the env variables are passed down from CLI. (edited)
lakstorm [03:20]
I am really confused about this whole thing.

That was a feature request by @jtopjian. We only do that if `--inherit-env` flag is provided.

**Comment by chenww** [comment]
how to add the flag --inherit-env via st2 UI?

**Comment by LindsayHill** [comment]
@chenww That `--inherit-env` action tells the CLI client to pass through the shell environment variables. That shell environment doesn't exist when you're running via the web UI.

If you want to pass something through as `env` in the web UI, then use the `env` parameter:

image



Also: Probably better to not post on very old, merged PRs. Hardly anyone will see those comments. Better to create a new issue, or for help, post to forum.stackstorm.com

**Comment by chenww** [comment]
@LindsayHill thanks for the reply. I thought env is from action runner, not related to web UI or not.

I checked API, https://api.stackstorm.com/api/v1/actions/#/actions_controller.post, same missing piece

**Pull request merged**